wgpu-core/src/instance.rs

```diff
@@ -298,6 +349,118 @@ impl<F: IdentityFilter<AdapterId>> Global<F> {
         adapters
     }
 
+    pub fn select_adapter(
+        &self,
+        desc: &AdapterInfo,
+        inputs: AdapterInputs<F::Input>,
+    ) -> Option<AdapterId> {
+        let instance = &self.instance;
+
+        match desc.backend {
+            #[cfg(any(
+                not(any(target_os = "ios", target_os = "macos")),
+                feature = "gfx-backend-vulkan"
+            ))]
+            Backend::Vulkan => {
+                if let Some(ref inst) = instance.vulkan {
+                    if let Some(id_vulkan) = inputs.find(Backend::Vulkan) {
+                        let adapter = inst
+                            .enumerate_adapters()
+                            .into_iter()
+                            .find(|adapter| desc == &adapter.info);
+
+                        if let Some(raw) = adapter {
+                            let adapter = Adapter { raw };
+                            log::info!("Adapter Vulkan {:?}", adapter.raw.info);
+                            let mut token = Token::root();
+                            return Some(
+                                backend::Vulkan::hub(self)
+                                    .adapters
+                                    .register_identity(id_vulkan, adapter, &mut token),
+                            );
+                        }
+                    }
+                }
+
+                None
+            }
+            #[cfg(any(target_os = "ios", target_os = "macos"))]
+            Backend::Metal => {
+                if let Some(id_metal) = inputs.find(Backend::Metal) {
+                    let adapter = instance
+                        .metal
+                        .enumerate_adapters()
+                        .into_iter()
+                        .find(|adapter| desc == &adapter.info);
```

Comment:
I'm slightly concerned that this approach wouldn't allow you to *really* select an adapter, i.e. in a case where there are 2 GPUs that are exactly the same. Current PR code would always select the first one.

Reply:
So we have to do it with `AdapterId`s then? Or is there some other way?

Reply:
One way could be to add something like an "array index" to the adapter info, so that one can select Nth adapter of the same (otherwise) info data.
Another approach is to go back to the Ids: return a bunch of Ids, let the user query their infos, and require the user to destroy unused Ids.

Reply:
As long as this "array index" would still require us using `register_identity` I think just returning the Ids is better.
I am going to return Ids; that's great: we made full circle again 😄.

Reply:
it could be the same as yours. Basically, when you obtained the HAL's adapter info list, you'd bump the "array index" for those infos that are exactly the same. Doing the opposite then, when an adapter is requested, would allow you to pick the correct adapter.

Btw, another thing you could do is return something that is dereffable to `AdapterInfo`. This way you wouldn't need to call `enumerate_adapters` again when a particular info is requested. This would make "array index" things unnecessary as well.

What do you think?
  1. return something dereffable to adapter info
  2. add "array index" to adapter info
  3. return IDs

Reply:
I was thinking that it maybe wouldn't be such a bad idea to put an `index` field in `AdapterInfo`, it would be useful for the user to be able to differentiate between equal GPUs too.
> Btw, another thing you could do is return something that is dereffable to `AdapterInfo`. This way you wouldn't need to call `enumerate_adapters` again when a particular info is requested. 

The `enumerate_adapters` you are referring to here is `instance.{BACKEND}.enumerate_adapters()` I assume?
I don't see how we could return something that would help us skip calling this again in `select_adapter` unless it's `gfx_hal::adapter::Adapter<B>`, which would put us back again at creating an enum abstracting over backends or `Box`ing them.
I think I'm not getting something here.

Reply:
> 1. return something dereffable to adapter info

Personally I would prefer using `AdapterId`s instead a new type that only helps to enumerate adapters.
> 2. add "array index" to adapter info

I just attempted doing this, as outlined above, I would like to do this in any case because I think this might be useful for the user to have in any case.
A problem I encountered though was `adapter_get_info`, there was no way for it to determine the correct `index` when building `AdapterInfo`.
This could be solved by either storing the index in `AdapterId`, or enumerate again through adapters to find out which index it is, but that would require to somehow be able to compare `gfx_hal::adapter::Adapter<B>`, which doesn't implement `Eq` or anything.
> 3. return IDs

Still the simplest of all solutions.

---

I encountered another potential problem: using an index, wouldn't be able to handle if an adapter is changing in the system, for example deactivating a GPU would make any form of identifier returned by `Global::enumerate_adapters()` invalid, unless it's an `AdapterId`, I assume.
I don't actually know how or if we are handling a case like that in gfx-hal or wgpu.

Also there is still the option to return an enum that abstracts over `gfx_hal::adapter::Adapter<B>`, which I know is something wgpu is trying to avoid.

Is there a possibility, to fix this issues somehow in gfx-hal? Would it be possible to return a unique identifier in `gfx_hal::adapter::Adapter<B>`?
I looked into it a bit, but my experience in this is really limited, I couldn't find anything clear-cut.

Personally I would prefer to make `AdapterInfo` a unique identifier, or stick with `AdapterId`, because thats a unique identifier too.

Reply:
To clarify, what I originally meant was having the actual adapter struct hidden by something that is dereffable to the adapter info. As you correctly noted, this would require an `enum`, so it's not the best.

You also raise a good point about how gfx-rs could make the adapter unique. I haven't done the proper investigation on it yet, and it seems quite far from the scope. gfx-rs returns a list of actual adapters, and it technically allows the wrappers to figure out how to identify any in that list. gfx-rs isn't made with the idea that you'd re-query the adapters and try to see which changed.

Using an "array index" is indeed prone to adapters changing, but I also think it's a bigger issue here. If the adapters have changed, and you are requesting something that you previously queried, many bets are off...

With that in mind, it indeed looks like we should be back to using the IDs.

Reply:
> With that in mind, it indeed looks like we should be back to using the IDs.

Alright, we are ready to merge then I assume? 🎉